### PR TITLE
Simplify the generated code for simple alternations

### PR DESCRIPTION
**Pull request by renatahodovan**
Simirarily to alternations over single literals, this change simplifies switch-like branches of alternations over single rules or literals to an index-based random selection from the list of alternatives.

### CODE REVIEW

File: grammarinator/process.py [hunk 1]
```diff
@@ -98,10 +98,21 @@ def __init__(self, idx, conditions):
...
+        # element is a list of literal values and rule ids if the alternation contains single literals
+        # and rule references only.
+        if not self.out_neighbours or any(len(alt.out_neighbours) != 1 or not isinstance(alt.out_neighbours[0], (LiteralNode, RuleNode)) for alt in self.out_neighbours):
+            return None, None, None
+
+        if all(isinstance(alt.out_neighbours[0], LiteralNode) for alt in self.out_neighbours):
+            return [repr(alt.out_neighbours[0].src) for alt in self.out_neighbours], None, None
+        if all(isinstance(alt.out_neighbours[0], RuleNode) for alt in self.out_neighbours):
+            return None, [alt.out_neighbours[0].id for alt in self.out_neighbours], None
+        return None, None, [(repr(alt.out_neighbours[0].src), None) if isinstance(alt.out_neighbours[0], LiteralNode) else (None, alt.out_neighbours[0].id) for alt in self.out_neighbours]
```
Comment (akosthekiss) [inline grammarinator/process.py:115]: `repr` should not get called on literals, neither at this line nor at some lines above. The result of this function will be used in jinja template(s) which may eventually support multiple languages, not only Python. And `repr` may add single or double quotes around `src`, which may not fit well with other programming languages. So, `repr` should happen on the Python template side. (And if and when another language gets supported, it should convert `src` to its language literal format as it sees fit.)

File: grammarinator/resources/codegen/GeneratorTemplate.py.jinja [hunk 2]
```diff
@@ -48,9 +48,17 @@ if self._max_depth >= {{ 0 if node.min == 1 else node.min_depth }}:
 
 {% macro processAlternationNode(node) %}
 choice = self._model.choice(current, {{ node.idx }}, [0 if [{{ node.min_depth | join(', ') }}][i] > self._max_depth else w for i, w in enumerate([{{ node.conditions | join(', ') }}])])
-{% set literal_alts = node.literal_alternatives() %}
-{% if literal_alts %}
-UnlexerRule(src=[{{ literal_alts | join(', ')}}][choice], parent=current)
+{% set simple_lits, simple_rules, mixed_alts = node.simple_alternatives() %}
+{% if simple_lits %}
+UnlexerRule(src=[{{ simple_lits | join(', ')}}][choice], parent=current)
```
Comment (akosthekiss) [inline grammarinator/resources/codegen/GeneratorTemplate.py.jinja:53]: Following on the comment on `repr`, this line may get a bit more complex. Something like:
```py
UnlexerRule(src=[{% for lit in simple_lits %}{{ lit | repr }}{% if not loop.last %}, {% endif %}{% endfor %}][choice], parent=current)
```

File: grammarinator/resources/codegen/GeneratorTemplate.py.jinja [hunk 3]
```diff
@@ -48,9 +48,17 @@ if self._max_depth >= {{ 0 if node.min == 1 else node.min_depth }}:
...
+{% elif simple_rules %}
+[{% for rule in simple_rules %}self.{{ rule }}{% if not loop.last %}, {% endif %}{% endfor %}][choice](parent=current)
+{% elif mixed_alts %}
+alt = [{% for alt in mixed_alts %}{% if alt[0] %}({{ alt[0] }}, {{ alt[1] }}){% else %}({{ alt[0] }}, self.{{ alt[1] }}){% endif %}{% if not loop.last %}, {% endif %}{% endfor %}][choice]
```
Comment (akosthekiss) [inline grammarinator/resources/codegen/GeneratorTemplate.py.jinja:57]: Similarly here. Moreover, as `repr` is not called during process, `alt[0]` may be an empty string, which is also falsy. So the `if alt[0]` test should be more specific to test for None. Something like:

```py
alt = [{% for alt in mixed_alts %}{% if alt[0] is None %}({{ alt[0] | repr }}, {{ alt[1] }}){% else %}({{ alt[0] }}, self.{{ alt[1] }}){% endif %}{% if not loop.last %}, {% endif %}{% endfor %}][choice]
```

**Review by akosthekiss — CHANGES_REQUESTED** [review]

File: grammarinator/process.py [hunk 4]
```diff
@@ -621,6 +632,7 @@ def __init__(self, lang, work_dir=None):
         env = Environment(trim_blocks=True,
                           lstrip_blocks=True,
                           keep_trailing_newline=False)
+        env.filters['repr'] = lambda s: repr(s)
```
Comment (akosthekiss) [inline grammarinator/process.py:635]: Well, it was me suggesting using the `repr` filter, which turned out not to exists as a built-in filter in Jinja. But there is, according to documentation, an alternative that should work without externally defining `repr`. As operator `%` is available in template expressions `"%r" % foo` shoud work out of the box.

But, here comes the but. The current way the template handles literals looks like as follows:

```py
{% macro processLiteralNode(node) %}
UnlexerRule(src='{{ node.src }}', parent=current)
{% endmacro %}
```

I.e., no "repr" or "%r" or anything like that is used, but `node.src` is simply surrounded by single quotes. I'm not completely sure why that works, why there is no need for any escaping, e.g., for a single quote literal. But if we want to maintain compatibility with the existing approach, then whenever we output a literal, we should use the same approach. (And if we find a better approach, it has to be applied at all places where literals are output.)

File: grammarinator/resources/codegen/GeneratorTemplate.py.jinja [hunk 5]
```diff
@@ -48,9 +48,17 @@ if self._max_depth >= {{ 0 if node.min == 1 else node.min_depth }}:
...
-{% set literal_alts = node.literal_alternatives() %}
-{% if literal_alts %}
-UnlexerRule(src=[{{ literal_alts | join(', ')}}][choice], parent=current)
+{% set simple_lits, simple_rules, mixed_alts = node.simple_alternatives() %}
+{% if simple_lits %}
+UnlexerRule(src=[{% for lit in simple_lits %}{{ lit | repr }}{% if not loop.last %}, {% endif %}{% endfor %}][choice], parent=current)
+{% elif simple_rules %}
+[{% for rule in simple_rules %}self.{{ rule }}{% if not loop.last %}, {% endif %}{% endfor %}][choice](parent=current)
+{% elif mixed_alts %}
+alt = [{% for alt in mixed_alts %}{% if alt[0] is None %}({{ alt[0] | repr }}, {{ alt[1] }}){% else %}({{ alt[0] }}, self.{{ alt[1] }}){% endif %}{% if not loop.last %}, {% endif %}{% endfor %}][choice]
```
Comment (akosthekiss) [inline grammarinator/resources/codegen/GeneratorTemplate.py.jinja:57]: Is `{% if alt[0] is None %}` right? Isn't the test negated? (Alternatively, `{% if alt[1] is None %}`?)

**Review by akosthekiss — CHANGES_REQUESTED** [review]

File: grammarinator/process.py [hunk 6]
```diff
@@ -98,10 +98,21 @@ def __init__(self, idx, conditions):
...
+        # element is a list of literal values and rule ids if the alternation contains single literals
+        # and rule references only.
+        if not self.out_neighbours or any(len(alt.out_neighbours) != 1 or not isinstance(alt.out_neighbours[0], (LiteralNode, RuleNode)) for alt in self.out_neighbours):
+            return None, None, None
+
+        if all(isinstance(alt.out_neighbours[0], LiteralNode) for alt in self.out_neighbours):
+            return [alt.out_neighbours[0].src for alt in self.out_neighbours], None, None
+        if all(isinstance(alt.out_neighbours[0], RuleNode) for alt in self.out_neighbours):
+            return None, [alt.out_neighbours[0].id for alt in self.out_neighbours], None
+        return None, None, [(alt.out_neighbours[0].src, None) if isinstance(alt.out_neighbours[0], LiteralNode) else (None, alt.out_neighbours[0].id) for alt in self.out_neighbours]
```
Comment (akosthekiss) [inline grammarinator/process.py:115]: I was thinking a bit about simplifying the values returned by this function - whether returning only two lists, i.e., a list of literals and a list of rules, could suffice. It could help get rid of that third list of tuples. What I came up with is below:

```py
def simple_alternatives(self):
      # Check if an alternation contains simple alternatives only (simple
      # literals or rule references), and return a 2-tuple. If the alternation
      # contains any non-simple alternatives, return None, None. If the
      # alternation contains simple literals only, the first element of the
      # tuple is a list of the literal values, while the second element is None.
      # If the alternation contains rule references only, the first element is
      # None, while the second element is a list of rule ids. If the alternation
      # contains both simple literals and rule references, then both elements of
      # the tuple are lists, which are of identical length, and exactly one of
      # them contains a non-None value at every index position.
      if not self.out_neighbours or any(len(alt.out_neighbours) != 1 or not isinstance(alt.out_neighbours[0], (LiteralNode, RuleNode)) for alt in self.out_neighbours):
          return None, None

      simple_lits = [alt.out_neighbours[0].src if isinstance(alt.out_neighbours[0], LiteralNode) else None for alt in self.out_neighbours]
      if all(lit is None for lit in simple_lits):
          simple_lits = None

      simple_rules = [alt.out_neighbours[0].id if isinstance(alt.out_neighbours[0], RuleNode) else None for alt in self.out_neighbours]
      if all(rule is None for rule in simple_rules):
          simple_rules = None

      return simple_lits, simple_rules
```

File: grammarinator/resources/codegen/GeneratorTemplate.py.jinja [hunk 7]
```diff
@@ -48,9 +48,17 @@ if self._max_depth >= {{ 0 if node.min == 1 else node.min_depth }}:
...
+{% if simple_lits %}
+UnlexerRule(src=[{% for lit in simple_lits %}'{{ lit }}'{% if not loop.last %}, {% endif %}{% endfor %}][choice], parent=current)
+{% elif simple_rules %}
+[{% for rule in simple_rules %}self.{{ rule }}{% if not loop.last %}, {% endif %}{% endfor %}][choice](parent=current)
+{% elif mixed_alts %}
+alt = [{% for alt in mixed_alts %}{% if alt[1] is none %}('{{ alt[0] }}', {{ alt[1] }}){% else %}({{ alt[0] }}, self.{{ alt[1] }}){% endif %}{% if not loop.last %}, {% endif %}{% endfor %}][choice]
+if alt[0] is not None:
+    UnlexerRule(src=alt[0], parent=current)
+else:
+    alt[1](parent=current)
```
Comment (akosthekiss) [inline grammarinator/resources/codegen/GeneratorTemplate.py.jinja:61]: This template could be adapted with just a bit of reordering:

```py
{% set simple_lits, simple_rules = node.simple_alternatives() %}
{% if simple_lits and simple_rules %}
src = [{% for lit in simple_lits %}{% if lit is not none %}'{{ lit }}'{% else %}None{% endif %}{% if not loop.last %}, {% endif %}{% endfor %}][choice]
rule = [{% for rule in simple_rules %}{% if rule is not none %}self.{{ rule }}{% else %}None{% endif %}{% if not loop.last %}, {% endif %}{% endfor %}][choice]
if src is not None:
    UnlexerRule(src=src, parent=current)
else:
    rule(parent=current)
{% if simple_lits %}
UnlexerRule(src=[{% for lit in simple_lits %}'{{ lit }}'{% if not loop.last %}, {% endif %}{% endfor %}][choice], parent=current)
{% elif simple_rules %}
[{% for rule in simple_rules %}self.{{ rule }}{% if not loop.last %}, {% endif %}{% endfor %}][choice](parent=current)
{% else %}
```

Let me know whether this approach is worth considering.

Comment (renatahodovan) [reply grammarinator/resources/codegen/GeneratorTemplate.py.jinja:61]: Thanks, it's a cleaner solution indeed.

**Review by akosthekiss — APPROVED** [review]
lgtm
